File: Dot examples/ASP Book Examples/Book Examples/Chapter17/CS/DBMovie.cs
using System;
using System.Text;
using Microsoft.SqlServer.Server;
using System.Data.SqlTypes;
using System.Runtime.InteropServices;
using System.IO;

[SqlUserDefinedType(Format.UserDefined, MaxByteSize = 512, IsByteOrdered = true)]
public class DBMovie : INullable, IBinarySerialize
{
    private bool _isNull;
    private string _title;
    private string _director;
    private decimal _boxOfficeTotals;

    public bool IsNull
    {
        get { return _isNull; }
    }

    public static DBMovie Null
    {
        get
        {
            DBMovie movie = new DBMovie();
            movie._isNull = true;
            return movie;
        }
    }

    public string Title
    {
        get { return _title; }
        set { _title = value; }
    }


    public string Director
    {
        get { return _director; }
        set { _director = value; }
    }

    [SqlFacet(Precision = 38, Scale = 2)]
    public decimal BoxOfficeTotals
    {
        get { return _boxOfficeTotals; }
        set { _boxOfficeTotals = value; }
    }


    [SqlMethod(OnNullCall = false)]
    public static DBMovie Parse(SqlString s)
    {
        if (s.IsNull)
            return Null;

        DBMovie movie = new DBMovie();
        string[] parts = s.Value.Split(new char[] { ',' });
        movie.Title = parts[0];
        movie.Director = parts[1];
        movie.BoxOfficeTotals = decimal.Parse(parts[2]);
        return movie;
    }

    public override string ToString()
    {
        if (this.IsNull)
            return "NULL";

        StringBuilder builder = new StringBuilder();
        builder.Append(_title);
        builder.Append(",");
        builder.Append(_director);
        builder.Append(",");
        builder.Append(_boxOfficeTotals.ToString());
        return builder.ToString();
    }

    public void Write(BinaryWriter w)
    {
        w.Write(_title);
        w.Write(_director);
        w.Write(_boxOfficeTotals);
    }

    public void Read(BinaryReader r)
    {
        _title = r.ReadString();
        _director = r.ReadString();
        _boxOfficeTotals = r.ReadDecimal();
    }

    public DBMovie()
    {
    }
}

File: Dot examples/ASP Book Examples/Book Examples/Chapter17/VB/DBMovie.vb
Imports System
Imports System.Text
Imports Microsoft.SqlServer.Server
Imports System.Data.SqlTypes
Imports System.Runtime.InteropServices
Imports System.IO

<SqlUserDefinedType(Format.UserDefined, MaxByteSize:=512, IsByteOrdered:=True)> _
Public Class DBMovie
    Implements INullable
    Implements IBinarySerialize

    Private _isNull As Boolean
    Private _title As String
    Private _director As String
    Private _boxOfficeTotals As Decimal

    Public ReadOnly Property IsNull() As Boolean Implements INullable.IsNull
        Get
            Return _isNull
        End Get
    End Property

    Public Shared ReadOnly Property Null() As DBMovie
        Get
            Dim movie As New DBMovie()
            movie._isNull = True
            Return movie
        End Get
    End Property

    Public Property Title() As String
        Get
            Return _title
        End Get
        Set(ByVal Value As String)
            _title = Value
        End Set
    End Property


    Public Property Director() As String
        Get
            Return _director
        End Get
        Set(ByVal Value As String)
            _director = Value
        End Set
    End Property

    <SqlFacet(Precision:=38, Scale:=2)> _
    Public Property BoxOfficeTotals() As Decimal
        Get
            Return _boxOfficeTotals
        End Get
        Set(ByVal Value As Decimal)
            _boxOfficeTotals = Value
        End Set
    End Property


    <SqlMethod(OnNullCall:=False)> _
    Public Shared Function Parse(ByVal s As SqlString) As DBMovie
        If (s.IsNull) Then
            Return Null
        End If
        Dim movie As New DBMovie()
        Dim parts() As String = s.Value.Split(New Char() {","c})
        movie.Title = parts(0)
        movie.Director = parts(1)
        movie.BoxOfficeTotals = Decimal.Parse(parts(2))
        Return movie
    End Function

    Public Overrides Function ToString() As String
        If Me.IsNull Then
            Return "NULL"
        End If
        Dim builder As New StringBuilder()
        builder.Append(_title)
        builder.Append(",")
        builder.Append(_director)
        builder.Append(",")
        builder.Append(_boxOfficeTotals.ToString())
        Return builder.ToString()
    End Function

    Public Sub Write(ByVal w As BinaryWriter) Implements IBinarySerialize.Write
        w.Write(_title)
        w.Write(_director)
        w.Write(_boxOfficeTotals)
    End Sub

    Public Sub Read(ByVal r As BinaryReader) Implements IBinarySerialize.Read
        _title = r.ReadString()
        _director = r.ReadString()
        _boxOfficeTotals = r.ReadDecimal()
    End Sub

    Public Sub New()
    End Sub
End Class

